### PR TITLE
Add feature map history

### PR DESCRIPTION
**Pull request by Hoytron**
## Proposed change

  HA displays historical data everywhere. Most of the time as timelines in line charts. But there is no option to show historical location data. So this PR add a feature to the standard `hui-map-card` and provides the option to show the history of a `sensor.<device-name>_geocoded_location` entity, which all android and iOS devices provide.

Example:
hui-map-card-history-feature

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml
type: map
history_entities:
  - entity: sensor.demo_paulus_geocoded_location
  - entity: sensor.demo_anne_therese_geocoded_location
hours_to_show: 12
```

## Additional information

This is my first change to the home-assistant frontend, I am happy with any improvements or suggestions for better code or feature integration. 

- Things to discuss: 
  - I added a range of default colors to the map, so different history paths are distinguishable. Please review the set of colors in https://github.com/Hoytron/frontend/blob/29c1d8fb54524d3f5cacfde48862c670292363d2/src/panels/lovelace/cards/hui-map-card.ts#L377
  - The line chart history function is used to retrieve and cache the historical location data. Maybe it is cleaner code to provide a overall historical cache for state attributes within the frontend
  - The source entity has to be a `_geocoded_location`, because its state changes independent from any zones. To use the device_tracker as data source, a mobile device needs to provide a sort of location entity with a location (latitude, longitude and maybe altitude) as state, so we can access its history. This requires changes in the backend and the mobile apps as far as I can see, and can be a future enhancement to this feature
  - The feature is not configurable with the visual editor for now. But its prepared to extend the visual editor easily.

- Link to documentation pull request: 
https://github.com/home-assistant/home-assistant.io/pull/12506

## Checklist

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

**Comment by homeassistant** [comment]
Hi @Hoytron, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;frontend#5331).

Once you do that we will be able to review and accept this pull request.

Thanks!

File: src/panels/lovelace/cards/types.ts [hunk 1]
```diff
@@ -139,6 +139,8 @@ export interface MapCardConfig extends LovelaceCardConfig {
   aspect_ratio?: string;
   default_zoom?: number;
   entities?: Array<EntityConfig | string>;
+  history_entities?: Array<EntityConfig | string>;
```
Comment (zsarnett) [inline src/panels/lovelace/cards/types.ts]: Is adding a new array type better or would it make more sense to add it the entity config in the entities array
```yaml
type: map
entities:
   - entity: location.entity
      show_history: true
```

File: src/panels/lovelace/editor/config-elements/hui-map-card-editor.ts [hunk 2]
```diff
@@ -67,6 +71,10 @@ export class HuiMapCardEditor extends LitElement implements LovelaceCardEditor {
     return this._config!.geo_location_sources || [];
   }
 
+  get _hours_to_show(): number {
+    return this._config!.hours_to_show || NaN;
```
Comment (zsarnett) [inline src/panels/lovelace/editor/config-elements/hui-map-card-editor.ts]: This should default to the default on the card. Probably 24 hours

File: src/panels/lovelace/cards/hui-map-card.ts [hunk 3]
```diff
@@ -329,8 +372,109 @@ class HuiMapCard extends LitElement implements LovelaceCard {
       }
     }
 
+    // DRAW history
+    if (config.history_entities && this._stateHistory?.timeline) {
+      const colors = [
```
Comment (zsarnett) [inline src/panels/lovelace/cards/hui-map-card.ts]: These colors should probably not be hardcoded. So that users can theme them. But if they do need to be in an array they should be in `src\data\map.ts` Which you will have to create

File: src/panels/lovelace/cards/hui-map-card.ts [hunk 4]
```diff
@@ -196,6 +231,7 @@ class HuiMapCard extends LitElement implements LovelaceCard {
 
   protected firstUpdated(changedProps: PropertyValues): void {
     super.firstUpdated(changedProps);
+    this._getStateHistory();
```
Comment (zsarnett) [inline src/panels/lovelace/cards/hui-map-card.ts]: I don't think you need this here, right? should just get run in updated

File: src/panels/lovelace/cards/hui-map-card.ts [hunk 5]
```diff
@@ -222,6 +258,9 @@ class HuiMapCard extends LitElement implements LovelaceCard {
 
   protected updated(changedProps: PropertyValues): void {
     if (changedProps.has("hass")) {
+      this._getStateHistory();
```
Comment (zsarnett) [inline src/panels/lovelace/cards/hui-map-card.ts]: We should only update this every minute. Like we do in the sensor card

File: src/panels/lovelace/cards/hui-map-card.ts (hunk 3 above)
Comment (bramkragten) [reply src/panels/lovelace/cards/hui-map-card.ts]: We have a list of colors here, we should use them.
<https://github.com/home-assistant/frontend/blob/dev/src/components/entity/ha-chart-base.js#L540>

File: src/panels/lovelace/cards/types.ts (hunk 1 above)
Comment (Hoytron) [reply src/panels/lovelace/cards/types.ts]: Wow, thanks for the fast review!
I tried to seperate this, to allow geocoded_location entities without a device_tracker entity and vice versa. 
To retrieve the required information the cache filters for _geocoded_location entities. So it is maybe confusing to provide another entity in the configuration than the entity is used to retrieve the information.

File: src/panels/lovelace/cards/hui-map-card.ts (hunk 3 above)
Comment (Hoytron) [reply src/panels/lovelace/cards/hui-map-card.ts]: > 
> 
> These colors should probably not be hardcoded. So that users can theme them. But if they do need to be in an array they should be in `src\data\map.ts` Which you will have to create

Is there a special way to provide theme colors, or just via the configuration file?

Comment (Hoytron) [reply src/panels/lovelace/cards/hui-map-card.ts]: > 
> 
> We have a list of colors here, we should use them.
> https://github.com/home-assistant/frontend/blob/dev/src/components/entity/ha-chart-base.js#L540

Nice. Maybe it is useful to move them to a separate file, to provided central color information?

File: src/panels/lovelace/cards/types.ts (hunk 1 above)
Comment (bramkragten) [reply src/panels/lovelace/cards/types.ts]: Why do you use the `geocoded_location` sensors and not the `device_tracker` history?

Comment (balloob) [reply src/panels/lovelace/cards/types.ts]: Bram is right, we should just query the history of the actual entity that we're showing.

Comment (Hoytron) [reply src/panels/lovelace/cards/types.ts]: I tried to do so. That was my first try and is its the best solution. 
But the attributes of an entity are only saved on state changes (as far as I know), and the state of a device_tracker ist unfortunatley bounded to zones. So the resulting path would only connect zones. 

>
> - The source entity has to be a _geocoded_location, because its state changes independent from any zones. To use the device_tracker as data source, a mobile device needs to provide a sort of location entity with a location (latitude, longitude and maybe altitude) as state, so we can access its history. This requires changes in the backend and the mobile apps as far as I can see, and can be a future enhancement to this feature

Comment (balloob) [reply src/panels/lovelace/cards/types.ts]: The device tracker is not bound to zones. It depends on how you have configured the app to report data. I'm surprised that the geocoded sensor is more reliable. 

We should not allow for different history entities. That' working around a problem with the data in our system. Instead, we should just make sure that our system records history correctly. Otherwise we're going to push this problem to every user that has to fix it, instead of us being able to just make it work.

I think that we should limit this feature to just 1 config option: `hours_to_show`. If not set, show no history. If set, show that time of history.

Comment (Hoytron) [reply src/panels/lovelace/cards/types.ts]: > 
> I think that we should limit this feature to just 1 config option: `hours_to_show`. If not set, show no history. If set, show that time of history.

I talked to @Santobert and he provides the backed change to realize your suggestion

File: src/data/history.ts [hunk 6]
```diff
@@ -55,7 +55,8 @@ export const fetchRecent = (
   entityId,
   startTime,
   endTime,
-  skipInitialState = false
+  skipInitialState = false,
+  includeLocation = false
```
Comment (balloob) [inline src/data/history.ts]: Pretty sure that this is not how we're going to call these options.

Probably be more likely to be called `fillStartPeriod` and `significantChangesOnly`


CC @Santobert

Comment (Santobert) [reply src/data/history.ts]: @balloob This is the backend pr: https://github.com/home-assistant/core/pull/33356

I've renamed `includeLocation` to `significantChangesOnly`. `skipInitialState` was already there and was not introduces by us, so I've decided to leave is as it is for now.

**Comment by Santobert** [comment]
Great :tada:  

grafik

**Comment by Hoytron** [comment]
> Great tada
> 
> These movements are almost 24 hours in the past, so the lines are a bit blurry:
> grafik
> 
> More recent movements look like this:
> grafik

Due to Covid19 and #stayathome the result is very good for all of us

**Comment by balloob** [comment]
Do we need blurry effects besides opacity changes ? Can you share a screenshot of one with just opacity? I think that it might look better.

**Comment by Hoytron** [comment]
The effect is only achieved by setting the opacity. The location path opcity is set linear from 100% opacity (most recent location) to 20% (oldest location to display).
 @Santobert surely means blurry, because he stayed at the same place the last hours and thus the locations in bigger distance (less opacity) are the oldest.

File: src/data/history.ts [hunk 7]
```diff
@@ -83,7 +83,10 @@ export const fetchDate = (
 ): Promise<HassEntity[][]> => {
   return hass.callApi(
     "GET",
-    `history/period/${startTime.toISOString()}?end_time=${endTime.toISOString()}`
+    `history/period/${startTime.toISOString()}
+      ?end_time=${endTime.toISOString()}
+      &significant_changes_only
```
Comment (balloob) [inline src/data/history.ts]: You are always adding this to the query string. That's not correct.

File: src/data/history.ts [hunk 8]
```diff
@@ -68,6 +69,9 @@ export const fetchRecent = (
   if (skipInitialState) {
     url += "&skip_initial_state";
   }
+  if (significantChangesOnly) {
+    url += "&significant_changes_only";
```
Comment (balloob) [inline src/data/history.ts]:
```suggestion
    url += `&significant_changes_only=${Number(significantChangesOnly)}`;
```

File: src/data/history.ts [hunk 9]
```diff
@@ -55,7 +55,8 @@ export const fetchRecent = (
   entityId,
   startTime,
   endTime,
-  skipInitialState = false
+  skipInitialState = false,
+  significantChangesOnly = true
```
Comment (balloob) [inline src/data/history.ts]: It's dangerous that we're encoding a set of defaults here on top of the defaults in the backend. Too late now for `skipIniitalState`, but let's not introduce another one with our own defaults.
```suggestion
  skipInitialState = false,
  significantChangesOnly?: boolean
```

File: src/data/history.ts [hunk 10]
```diff
@@ -68,6 +69,9 @@ export const fetchRecent = (
   if (skipInitialState) {
     url += "&skip_initial_state";
   }
+  if (significantChangesOnly) {
+    url += "&significant_changes_only";
+  }
```
Comment (balloob) [inline src/data/history.ts]:
```suggestion
  if (significantChangesOnly !== undefined) {
    url += `&significant_changes_only=${Number(significantChangesOnly)}`;
  }
```

File: src/panels/lovelace/cards/hui-map-card.ts [hunk 11]
```diff
@@ -218,10 +249,12 @@ class HuiMapCard extends LitElement implements LovelaceCard {
...
         : (root.style.paddingBottom = "100%");
+
+    this._date = new Date();
   }
 
   protected updated(changedProps: PropertyValues): void {
-    if (changedProps.has("hass")) {
+    if (changedProps.has("hass") || changedProps.has("_history")) {
       this._drawEntities();
       this._fitMap();
```
Comment (balloob) [inline src/panels/lovelace/cards/hui-map-card.ts]: This is existing code, but isn't it a bad experience that we keep fitting the map on a state change? It won't allow users to explore data, especially now that we're showing historic data.

File: src/panels/lovelace/cards/hui-map-card.ts [hunk 12]
```diff
@@ -453,6 +569,29 @@ class HuiMapCard extends LitElement implements LovelaceCard {
...
+      startTime,
+      endTime
+    );
+
+    if (stateHistory.length < 1) {
+      return;
+    }
+
+    this._history = stateHistory;
+    this._date = new Date();
```
Comment (balloob) [inline src/panels/lovelace/cards/hui-map-card.ts]: You need to set this before you are requesting data, otherwise if 5 state changes happen while you are fetching data, you're still fetching the history 5 times.

File: src/panels/lovelace/cards/hui-map-card.ts [hunk 13]
```diff
@@ -453,6 +569,29 @@ class HuiMapCard extends LitElement implements LovelaceCard {
     }
   }
 
+  private async _getHistory(): Promise<void> {
+    const entityIds = this._configEntities
+      ?.map((entity) => entity.entity)
```
Comment (balloob) [inline src/panels/lovelace/cards/hui-map-card.ts]: This variable is always there ? If it's not , we shouldn't request the history of all entities over a period of time. In that case we shouldn't fetch anything.

File: src/panels/lovelace/cards/hui-map-card.ts [hunk 14]
```diff
@@ -453,6 +569,29 @@ class HuiMapCard extends LitElement implements LovelaceCard {
...
+    const endTime = new Date();
+    const startTime = new Date();
+    startTime.setHours(endTime.getHours() - this._config!.hours_to_show!);
+
+    const stateHistory = await fetchRecent(
+      this.hass,
+      entityIds,
+      startTime,
+      endTime
+    );
```
Comment (balloob) [inline src/panels/lovelace/cards/hui-map-card.ts]: You're not using skipInitialState or include non-significant states.

File: src/panels/lovelace/cards/hui-map-card.ts (hunk 11 above)
Comment (zsarnett) [reply src/panels/lovelace/cards/hui-map-card.ts]: https://developers.google.com/maps/documentation/javascript/examples/control-custom

Probably for another PR though

File: src/panels/lovelace/cards/hui-map-card.ts [hunk 15]
```diff
@@ -283,16 +325,26 @@ class HuiMapCard extends LitElement implements LovelaceCard {
...
-    );
+    const bounds = this.Leaflet.featureGroup(this._mapItems).getBounds();
     this._leafletMap.fitBounds(bounds.pad(0.5));
 
     if (zoom && this._leafletMap.getZoom() > zoom) {
       this._leafletMap.setZoom(zoom);
     }
   }
 
+  private _getColor(name: string) {
```
Comment (balloob) [inline src/panels/lovelace/cards/hui-map-card.ts]: This is only called with entityId, and so parameter should be entityId.

File: src/panels/lovelace/cards/hui-map-card.ts [hunk 16]
```diff
@@ -283,16 +325,26 @@ class HuiMapCard extends LitElement implements LovelaceCard {
...
+    let color;
+    if (this._colorDict[name]) {
+      color = this._colorDict[name];
+    } else {
```
Comment (balloob) [inline src/panels/lovelace/cards/hui-map-card.ts]: This is not possible since we use entity IDs in this function only.

File: src/panels/lovelace/cards/hui-map-card.ts [hunk 17]
```diff
@@ -231,6 +264,15 @@ class HuiMapCard extends LitElement implements LovelaceCard {
     ) {
       this.updateMap(changedProps.get("_config") as MapCardConfig);
     }
+
+    if (this._config!.hours_to_show) {
```
Comment (balloob) [inline src/panels/lovelace/cards/hui-map-card.ts]:
```suggestion
    if (this._config!.hours_to_show && this._configEntities?.length) {
```

File: src/panels/lovelace/entity-rows/types.ts [hunk 18]
```diff
@@ -7,6 +7,7 @@ export interface EntityConfig {
   name?: string;
   icon?: string;
   image?: string;
+  color?: string;
```
Comment (balloob) [inline src/panels/lovelace/entity-rows/types.ts]: Unrelated change?

File: src/panels/lovelace/cards/hui-map-card.ts [hunk 19]
```diff
@@ -329,6 +386,63 @@ class HuiMapCard extends LitElement implements LovelaceCard {
...
+        ) as LatLngTuple[];
+
+        // DRAW HISTORY
+        for (
+          let markerIndex = 0;
+          markerIndex < path.length - 1;
+          markerIndex++
+        ) {
+          const opacityStep = 1 / path.length;
+          const opacity = 2 * opacityStep + markerIndex * opacityStep;
```
Comment (balloob) [inline src/panels/lovelace/cards/hui-map-card.ts]: If we simplify the above equation, you get:
```suggestion
          const opacity = (markerIndex + 2) * opacityStep;
```
This results in an opacity above 1?

Comment (Hoytron) [reply src/panels/lovelace/cards/hui-map-card.ts]: Actually not, because the loop condition is `markerIndex < path.length - 1 `. (I do not draw the dot underneath the current location). 
Same code here:  https://js.do/code/423192

But you are right, somehow I missed to start at 20% opacity. 
New code is be: https://js.do/code/423195

File: src/panels/lovelace/cards/hui-map-card.ts (hunk 11 above)
Comment (bramkragten) [reply src/panels/lovelace/cards/hui-map-card.ts]: @zsarnett this is not google maps

Comment (Hoytron) [reply src/panels/lovelace/cards/hui-map-card.ts]: balloob is right. We center the map on every hass change (and now history change), to fit all markers. But this is bad user experience, when user want to explore the historic data. 

In an extra PR I can add a 'center map' button to refit the map. And a inactivity tracker to refit the map would be useful, too.

File: src/panels/lovelace/cards/hui-map-card.ts (hunk 16 above)
Comment (Hoytron) [reply src/panels/lovelace/cards/hui-map-card.ts]: @balloob What do you mean by this is not possible? 
The dict is to ensure same entitys get the same color for marker and path. Are entity IDs not uniqe?

File: src/panels/lovelace/cards/hui-map-card.ts (hunk 15 above)
Comment (Hoytron) [reply src/panels/lovelace/cards/hui-map-card.ts]: I tried to keep this function more generic. input a name and if you request a color for the same name again, you get the same color. Independent what kind of name it is. 
But I have changed it.

File: src/panels/lovelace/entity-rows/types.ts (hunk 18 above)
Comment (Hoytron) [reply src/panels/lovelace/entity-rows/types.ts]: User was able to define own colors via yml configuration. 
Removed it, to keep focus on the feature itself.

File: src/panels/lovelace/cards/hui-map-card.ts (hunk 16 above)
Comment (balloob) [reply src/panels/lovelace/cards/hui-map-card.ts]: My bad. I didn't realize it was a caching function. This is ok.

File: src/data/history.ts (hunk 7 above)
Comment (balloob) [reply src/data/history.ts]: Why was this resolved? We shouldn't make this change as the backend is backwards compatible.

Comment (Hoytron) [reply src/data/history.ts]: You're right. my fault. I missed this one.

File: src/panels/lovelace/cards/hui-map-card.ts [hunk 20]
```diff
@@ -218,10 +249,12 @@ class HuiMapCard extends LitElement implements LovelaceCard {
       ratio && ratio.w > 0 && ratio.h > 0
         ? `${((100 * ratio.h) / ratio.w).toFixed(2)}%`
         : (root.style.paddingBottom = "100%");
+
+    this._date = new Date();
```
Comment (balloob) [inline src/panels/lovelace/cards/hui-map-card.ts]: We should do this inside `setConfig`. We should also reset the cached history.

Right now we will re-fetch history if config has changed into a config that shows history. This means that you cannot set history to 0.

**Review by balloob — APPROVED** [review]
Looks great! Ok to merge when we added code to clean up the history when no longer used / entities changed in config.

File: src/panels/lovelace/cards/hui-map-card.ts (hunk 20 above)
Comment (Hoytron) [reply src/panels/lovelace/cards/hui-map-card.ts]: We don't need it anymore. When we change config then `update()` is triggert and `if changedProps.has("_config")` then we fetch the history and while this we update the Date attribute.

Comment (balloob) [reply src/panels/lovelace/cards/hui-map-card.ts]: If new config has `hours_to_show = 0` it won't update the history nor clear it.

Comment (bramkragten) [reply src/panels/lovelace/cards/hui-map-card.ts]: But `_history` is never cleared.

Comment (bramkragten) [reply src/panels/lovelace/cards/hui-map-card.ts]: I'm too slow...

Comment (Hoytron) [reply src/panels/lovelace/cards/hui-map-card.ts]: I've added the history cleanup just now. 

> We don't need it anymore. When we change config then update() is triggert and if  changedProps.has("_config") then we fetch the history and while this we update the Date attribute

This was meant to the line of code above. Sorry for unclear statement

**Review by balloob — APPROVED** [review]
